src/Cassandra.Tests/RequestExecutionTests.cs

```diff
@@ -0,0 +1,184 @@
+﻿//
+//       Copyright DataStax, Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Net;
+
+using Cassandra.Requests;
+using Cassandra.Responses;
+using Cassandra.Serialization;
+
+using Moq;
+
+using NUnit.Framework;
+
+namespace Cassandra.Tests
+{
+    [TestFixture]
+    public class RequestExecutionTests
+    {
+        [Test, TestCase(true), TestCase(false)]
+        public void Should_ThrowException_When_NoValidHosts(bool currentHostRetry)
+        {
+            var mockSession = Mock.Of<IInternalSession>();
+            var mockRequest = Mock.Of<IRequest>();
+            var mockRequestExecution = Mock.Of<IRequestHandler>();
+            Mock.Get(mockRequestExecution)
+                .Setup(m => m.GetNextValidHost(It.IsAny<Dictionary<IPEndPoint, Exception>>()))
+                .Throws(new NoHostAvailableException(new Dictionary<IPEndPoint, Exception>()));
+            var sut = new ProxyRequestExecution(mockRequestExecution, mockSession, mockRequest);
+
+            Assert.Throws<NoHostAvailableException>(() => sut.Start(currentHostRetry));
+        }
+
+        [Test, TestCase(true), TestCase(false)]
+        public void Should_NotThrowException_When_AValidHostIsObtained(bool currentHostRetry)
+        {
+            var mockSession = Mock.Of<IInternalSession>();
+            var mockRequest = Mock.Of<IRequest>();
+            var mockRequestExecution = Mock.Of<IRequestHandler>();
+            var host = new Host(
+                new IPEndPoint(IPAddress.Parse("127.0.0.1"), 9047),
+                new ConstantReconnectionPolicy(1));
+            var validHost = ValidHost.New(
+                host,
+                HostDistance.Local);
+            Mock.Get(mockRequestExecution)
+                .SetupSequence(m => m.GetNextValidHost(It.IsAny<Dictionary<IPEndPoint, Exception>>()))
+                .Returns(validHost)
+                .Returns(null);
```

Comment:
The caller of this method doesn't expect a `null` value, causing an infinite recursion.
The second time it should throw a `NoHostAvailableException` instance.

Reply:
Hmm that actually brings up an interesting question, what if a user provides a custom load balancing policy to the driver that returns `null` instead of throwing an exception? I should probably add something to check that scenario and treat `null` the same way as thrown exceptions, right?

Reply:
Its a good question... I think `GetNextValidHost()` already validates for `null`, so callers should not expect `null` values. Its only a matter of the mock that's behaving in a forbidden way.

Reply:
Ah silly me you're right. Changed it!